### PR TITLE
SOI simulation files

### PR DESCRIPTION
**Pull request by YagmurSimsekk**

### CODE REVIEW

File: R/SimulateSOI.R [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+#'
+#' Generate time-series with the self-organised instability (SOI) model.
+#' Implements a K-leap method for accelerating stochastic simulation.
+#'
+#' @param N Integer: number of species
+#' @param I Integer: community size, number of available sites (individuals)
+#' @param A a interaction matrix of dimension NxN
+#' @param com a vector inditicating initial community abundances.
+#' If NULL (default), based on migration rates.
+#' @param tend Integer : number of timepoints to be returned in the time series
```
Comment (antagomir) [inline R/SimulateSOI.R:11]: remove unnecessary space before ":"

File: R/SimulateSOI.R [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+            # Last columns: the competition / negative interaction jumps:
+            trans_mat <- cbind(
+                rbind(diag(1, ncol = N, nrow = N), rep(-1, times = N)),
+                rbind(diag(-1, ncol = N, nrow = N), rep(1, times = N))
+            )
+
+            # Interaction:
+            # happens when happens when sp_stronger interacts more strongle
+            # with sp_weaker, than sp_weaker with sp_stronger
+
```
Comment (antagomir) [inline R/SimulateSOI.R:118]: remove empty lines like this

File: R/SimulateSOI.R [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+            pos_inter_rates <- matrix(0, nrow = N, ncol = N)
+            neg_inter_rates <- pos_inter_rates # copy to initialise
+
+            # pos interaction and immigration
+            # AND sp_stronger interacts positively with sp_weaker
+            for(stronger in 1:N){
```
Comment (antagomir) [inline R/SimulateSOI.R:124]: I think you're supposed to use `seq_len(N)` here instead?

File: R/SimulateSOI.R [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+                weaker <- which(A[stronger,] > A[,stronger] & A[,stronger] >= 0)
+            if(length(weaker) > 0){
+                rate <- A[stronger, weaker] + A[weaker, stronger]
+                pos_inter_rates[stronger,weaker] <- rate
+                }
+            }
```
Comment (antagomir) [inline R/SimulateSOI.R:130]: Something strange going on here with the indentation?

File: R/SimulateSOI.R (hunk 3 above)
Comment (antagomir) [reply R/SimulateSOI.R:124]: and in other similar places

File: R/SimulateSOI.R [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+                for(weaker in weaker_vec){
+                    jump <- rep(0, times = N+1)
+                    jump[stronger] <- 1
+                    jump[weaker] <- -1
+                    trans_mat <- cbind(trans_mat, jump)
+                    }
+                }
+            }
+
+            # INITIAL PROPENSITIES
```
Comment (antagomir) [inline R/SimulateSOI.R:148]: Start comment from the same position than subsequent line

File: R/SimulateSOI.R [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+            weaker_vec <- which(A[stronger,] > A[,stronger] & A[,stronger] < 0)
+                if(length(weaker_vec) > 0){
+                rate <- A[stronger, weaker_vec] - A[weaker_vec, stronger]
+                neg_inter_rates[stronger, weaker_vec] <- rate
+                for(weaker in weaker_vec){
+                    jump <- rep(0, times = N+1)
+                    jump[stronger] <- 1
+                    jump[weaker] <- -1
+                    trans_mat <- cbind(trans_mat, jump)
+                    }
```
Comment (antagomir) [inline R/SimulateSOI.R:144]: check indentation

File: R/SimulateSOI.R [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+                }
+            }
+
+            # INITIAL PROPENSITIES
+        propensities <- update_propensities(I, counts, death_rates, migr_rates,
+                                            pos_inter_rates, neg_inter_rates)
+
+            # K-LEAPS SIMULATION
+            sys_t <- seq(from = 0, to = tend, length.out = tend)
+
```
Comment (antagomir) [inline R/SimulateSOI.R:154]: Why we have indentation here suddenly changing? not good for readability of the code

File: R/SimulateSOI.R [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+            series <- matrix(0, nrow = N, ncol = tend)
+            colnames(series) <- paste0("t", 1:tend)
+            rownames(series) <- paste0("sp_", 1:N)
+
+            while(series_t <= tend){
+
+                c = sum(propensities)
+                p = propensities/c
+                tau = rgamma(n = 1, shape = k, scale = 1/c)
```
Comment (antagomir) [inline R/SimulateSOI.R:169]: should replace "=" with "<-"

File: R/SimulateSOI.R [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+                series_t <- series_t +1
+                index <- which(sys_t >= sample_t & sys_t < current_t)
+                sample_t <- sys_t[index][length(index)]
+            }
+
+                # which reactions allowed?
+                reactionsToFire <- as.vector(rmultinom(n= 1, size= k, prob = p))
+                # update counts with allowed transitions (reactions)
+                counts <- sapply(counts + trans_mat%*%reactionsToFire,
+                            FUN = function(x){x = max(0,x)})
```
Comment (antagomir) [inline R/SimulateSOI.R:190]: check indentation

File: R/SimulateSOI.R [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+        propensities <- update_propensities(I, counts, death_rates, migr_rates,
+                                            pos_inter_rates, neg_inter_rates)
+            }
+            if(norm){
+                series <- t(t(series)/colSums(series))
+            }
+
+            series
+            SOI <- SummarizedExperiment(assays = list(counts= series))
+            return(SOI)
```
Comment (antagomir) [inline R/SimulateSOI.R:201]: check indetnation

File: R/SimulateSOI.R [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+        # update propensities with updated counts
+        propensities <- update_propensities(I, counts, death_rates, migr_rates,
+                                            pos_inter_rates, neg_inter_rates)
+            }
+            if(norm){
+                series <- t(t(series)/colSums(series))
+            }
+
+            series
+            SOI <- SummarizedExperiment(assays = list(counts= series))
```
Comment (antagomir) [inline R/SimulateSOI.R:200]: remove space after "counts="

File: R/SimulateSOI.R [hunk 12]
```diff
@@ -0,0 +1,246 @@
...
+    stdev = 1, # sd normal distribution
+    s = 0.1 # scaling parameter, default: 0.1/max(A)
+){
+    # Nominal Interspecific Interaction matrix N
+    N <- matrix(
+        data = rnorm(n^2, mean = 0, sd = stdev),
+        nrow = n,
+        ncol = n
+    )
+    #diag(N) <- 0
```
Comment (antagomir) [inline R/SimulateSOI.R:216]: can remove this line

**Review by antagomir — CHANGES_REQUESTED** [review]

File: R/SimulateSOI.R (hunk 7 above)
Comment (YagmurSimsekk) [reply R/SimulateSOI.R:154]: It was because I avoid exceeding character length limit in a line but now I fixed it

File: NAMESPACE [hunk 13]
```diff
@@ -1,6 +1,11 @@
 # Generated by roxygen2: do not edit by hand
 
+export(dxdt)
+export(powerlawA)
 export(simulateGLV)
+export(simulateSOI)
+export(update_propensities)
```
Comment (antagomir) [inline NAMESPACE:7]: In our coding style, this should be something like `updatePropensities`.

Let's be consistent.

File: NAMESPACE [hunk 14]
```diff
@@ -1,6 +1,11 @@
 # Generated by roxygen2: do not edit by hand
 
+export(dxdt)
```
Comment (antagomir) [inline NAMESPACE:3]: what does this do? Can we give name such as `calculatedxdt`. If this does not have to be exported, then this does not matter and better to keep the name as is. If this must be expoerted, the name should be more descriptive.

File: R/SimulateSOI.R [hunk 15]
```diff
@@ -0,0 +1,222 @@
...
+    min(
+      ceiling(d*pl[i]/mean(pl)),
+      n
+    )
+  })
+  G <- matrix(0, nrow = n, ncol = n)
+  for(i in seq_len(n)){
+    index <- sample(x = seq_len(n), size = h[i])
+    G[index, i] <- 1
+  }
```
Comment (antagomir) [inline R/SimulateSOI.R:62]: This does not follow the 4-space indentation required by Bioconductor. Is there no warning about this in BiocCheck()?

Comment (antagomir) [reply R/SimulateSOI.R:62]: Same for rest of the code

File: R/simulateGLV.R [hunk 16]
```diff
@@ -24,53 +24,98 @@
...
+    G <- matrix(0, nrow = n, ncol = n)
+    for(i in seq_len(n)){
+        index <- sample(x = seq_len(n), size = h[i])
+        G[index, i] <- 1
+    }
+    A <- N %*% H * G
+    A <- A*s/max(A)
+    diag(A) <- -1
+    colnames(A) <- seq_len(n)
+    rownames(A) <- seq_len(n)
```
Comment (antagomir) [inline R/simulateGLV.R:95]: why duplicated?

**Review by antagomir — CHANGES_REQUESTED** [review]

File: R/simulateGLV.R [hunk 17]
```diff
@@ -24,53 +24,98 @@
...
+    n, # number of species
+    alpha, # power-law distribution parameter
+    stdev = 1, # sd normal distribution
+    s = 0.1 # scaling parameter, default: 0.1/max(A)
+){
+    # Nominal Interspecific Interaction matrix N
+    N <- matrix(
+        data = rnorm(n^2, mean = 0, sd = stdev),
+        nrow = n,
+        ncol = n
```
Comment (YagmurSimsekk) [inline R/simulateGLV.R:69]: i assumed that because it is nxn matrix so the names must be equal as well ?

File: R/simulateGLV.R (hunk 16 above)
Comment (antagomir) [reply R/simulateGLV.R:95]: ah this is ok I checked wrong

File: R/SimulateSOI.R (hunk 15 above)
Comment (YagmurSimsekk) [reply R/SimulateSOI.R:62]: Sorry, i copy pasted to new file just before uploading, it must be because of that. Now it is corrected and checked

**Pull request merged**